### PR TITLE
Support an access_token coming in a Cookie

### LINKED ISSUE #474: Support an access_token coming in a Cookie
Since we switched to Token Management, `access_token`s can only be properly validated when they come in the `Authorization` header. 

A request like this was previously working (1.0.RC8.2) but does not work any longer:

```
GET /stars/pollux
Host: galaxies.com

Cookie: access_token=eyJhbGciOiJIUzI1NiIsI.eyJpc3MiOiJodHRwczotcGxlL.mFrs3Zo8eaSNcxiNfvRh9dqKP4F1cB;
```

The issue is in `DefaultJwtAccountResolverr`

### PR DESCRIPTION
**Pull request by jarias**
https://github.com/stormpath/stormpath-framework-spec/blob/master/cookie-authentication.md

Fixes #474

### CODE REVIEW

**Comment by mraible** [comment]
As far as I can tell, this is failing because there's 60 keys in `web.stormpath.properties` that aren't in the spec. The expectation is that there's 66 keys. Did you remove 6 keys from `web.stormpath.properties`? If so, you should be able to easily update the test.

```
Failed tests: verifyPropertiesInDefaultAreInSpec(com.stormpath.sdk.servlet.config.SpecConfigVersusWebPropertiesTest): 
Missing keys in spec config: [...] expected [66] but found [60]
```

**Comment by jarias** [comment]
@mraible actually the issue was a 2 part, first a bad merge and I duplicate some properties, the second is that a couple of properties went away. I will update https://github.com/stormpath/stormpath-sdk-java/wiki/1.0-Configuration-Changes-&-Additions-Guide#not-in-specification

File: extensions/servlet/src/main/java/com/stormpath/sdk/servlet/config/CookieProperties.java [hunk 1]
```diff
@@ -0,0 +1,20 @@
+package com.stormpath.sdk.servlet.config;
+
+/**
+ * @since 1.0.RC10
+ */
+public interface CookieProperties {
+    String getCookieName();
```
Comment (mrioan) [inline extensions/servlet/src/main/java/com/stormpath/sdk/servlet/config/CookieProperties.java]: Please add a line break here

File: extensions/servlet/src/main/java/com/stormpath/sdk/servlet/config/impl/AbstractCookieConfig.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+import com.stormpath.sdk.lang.Assert;
+import com.stormpath.sdk.servlet.config.CookieConfig;
+import com.stormpath.sdk.servlet.config.CookieProperties;
+
+/**
+ * @since 1.0.RC10
+ */
+public abstract class AbstractCookieConfig implements CookieConfig {
+    protected static final String NAME = "name";
```
Comment (mrioan) [inline extensions/servlet/src/main/java/com/stormpath/sdk/servlet/config/impl/AbstractCookieConfig.java]: Please add a line break here

File: extensions/servlet/src/main/java/com/stormpath/sdk/servlet/config/impl/AccessTokenCookieConfig.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
+package com.stormpath.sdk.servlet.config.impl;
+
+import com.stormpath.sdk.servlet.config.CookieProperties;
+
+/**
+ * @since 1.0.RC10
+ */
+public class AccessTokenCookieConfig extends AbstractCookieConfig {
+    private static final String CONFIG_PREFIX = "stormpath.web.accessTokenCookie";
```
Comment (mrioan) [inline extensions/servlet/src/main/java/com/stormpath/sdk/servlet/config/impl/AccessTokenCookieConfig.java]: * Please add a line break here
* @since 1.0.0

File: extensions/servlet/src/main/java/com/stormpath/sdk/servlet/config/impl/RefreshTokenCookieConfig.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
+package com.stormpath.sdk.servlet.config.impl;
+
+import com.stormpath.sdk.servlet.config.CookieProperties;
+
+/**
+ * @since 1.0.RC10
+ */
+public class RefreshTokenCookieConfig extends AbstractCookieConfig {
+    private static final String CONFIG_PREFIX = "stormpath.web.refreshTokenCookie";
```
Comment (mrioan) [inline extensions/servlet/src/main/java/com/stormpath/sdk/servlet/config/impl/RefreshTokenCookieConfig.java]: * Please add a line break here
* @since 1.0.0

File: extensions/servlet/src/main/java/com/stormpath/sdk/servlet/config/impl/AbstractCookieConfig.java [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    public AbstractCookieConfig(ConfigReader configReader) {
+        Assert.notNull(configReader);
+        this.name = configReader.getString(configKeyFor(NAME));
+        Assert.hasText(this.name, NAME + " cannot be null or empty.");
+        this.comment = configReader.getString(configKeyFor(COMMENT));
+        this.domain = configReader.getString(configKeyFor(DOMAIN));
+        this.path = configReader.getString(configKeyFor(PATH));
+        this.secure = configReader.getBoolean(configKeyFor(SECURE));
+        this.httpOnly = configReader.getBoolean(configKeyFor(HTTP_ONLY));
+        this.maxAge = 1;
```
Comment (mrioan) [inline extensions/servlet/src/main/java/com/stormpath/sdk/servlet/config/impl/AbstractCookieConfig.java]: In `AccountCookieConfig` we had: `this.maxAge = Math.max(-1, configReader.getInt(MAX_AGE));`. Why is this changed to `1` here?

File: extensions/servlet/src/main/java/com/stormpath/sdk/servlet/config/impl/DefaultConfig.java [hunk 6]
```diff
@@ -81,19 +74,8 @@ public DefaultConfig(final ServletContext servletContext, Map<String, String> co
         this.CFG = new ExpressionConfigReader(servletContext, this.props);
         this.SINGLETONS = new LinkedHashMap<String, Object>();
 
-        this.ACCOUNT_COOKIE_CONFIG = new AccountCookieConfig(CFG);
-
-        long accountJwtTtl = CFG.getLong(ACCOUNT_JWT_TTL);
-        Assert.isTrue(accountJwtTtl > 0, ACCOUNT_JWT_TTL + " value must be a positive long.");
```
Comment (mrioan) [inline extensions/servlet/src/main/java/com/stormpath/sdk/servlet/config/impl/DefaultConfig.java]: I see this logic has not been moved anywhere, just removed. What is the rationale for this change?

File: extensions/servlet/src/main/java/com/stormpath/sdk/servlet/config/impl/DefaultConfig.java [hunk 7]
```diff
@@ -81,19 +74,8 @@ public DefaultConfig(final ServletContext servletContext, Map<String, String> co
...
-
-        int accountCookieMaxAge = this.ACCOUNT_COOKIE_CONFIG.getMaxAge();
-
-        if (accountCookieMaxAge > 0 && accountCookieMaxAge > accountJwtTtl) {
```
Comment (mrioan) [inline extensions/servlet/src/main/java/com/stormpath/sdk/servlet/config/impl/DefaultConfig.java]: I see this logic has not been moved anywhere, just removed. What is the rationale for this change? We do not have an account jwt's ttl any longer but we do have the access token TTL and refresh token TTL.

File: extensions/servlet/src/main/java/com/stormpath/sdk/servlet/filter/account/CookieAuthenticationResultSaver.java [hunk 8]
```diff
@@ -176,8 +174,20 @@ public boolean isSecure() {
...
+    private int getCookieMaxAge(CookieConfig cookieConfig) {
+        OAuthPolicy oauthPolicy = application.getOAuthPolicy();
+
+        if (cookieConfig instanceof AccessTokenCookieConfig) {
+            return Period.parse(oauthPolicy.getAccessTokenTtl()).toStandardSeconds().getSeconds();
+        } else if (cookieConfig instanceof RefreshTokenCookieConfig) {
+            return Period.parse(oauthPolicy.getRefreshTokenTtl()).toStandardSeconds().getSeconds();
+        }
+
+        return 259200;
```
Comment (mrioan) [inline extensions/servlet/src/main/java/com/stormpath/sdk/servlet/filter/account/CookieAuthenticationResultSaver.java]: Why is this value hardcoded here? If it is part of the spec please move it to a final property in this class and add document it by referencing that document. Thanks

Comment (mrioan) [inline extensions/servlet/src/main/java/com/stormpath/sdk/servlet/filter/account/CookieAuthenticationResultSaver.java]: Why is this value hardcoded here? If it is part of the spec please move it to a final property in this class and add document it by referencing that document. Thanks

File: extensions/servlet/src/main/java/com/stormpath/sdk/servlet/oauth/OAuthTokenResolver.java [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+/**
+ * An OAuthTokenResolver can inspect a {@link javax.servlet.ServletRequest ServletRequest} and return either
+ * an Access Token or Refresh Token that may be associated with the request due to a previous
+ * authentication.
+ *
+ * @since 1.0.RC10
+ */
+public interface OAuthTokenResolver {
+    String REQUEST_ATTR_NAME = AccessTokenResolver.class.getName();
```
Comment (mrioan) [inline extensions/servlet/src/main/java/com/stormpath/sdk/servlet/oauth/OAuthTokenResolver.java]: * Please add a line break here
* @since 1.0.0

File: extensions/servlet/src/main/java/com/stormpath/sdk/servlet/oauth/OAuthTokenResolver.java [hunk 10]
```diff
@@ -0,0 +1,61 @@
+package com.stormpath.sdk.servlet.oauth;
+
+import com.stormpath.sdk.servlet.oauth.impl.AccessTokenResolver;
+
+import javax.servlet.ServletRequest;
+
+/**
+ * An OAuthTokenResolver can inspect a {@link javax.servlet.ServletRequest ServletRequest} and return either
+ * an Access Token or Refresh Token that may be associated with the request due to a previous
+ * authentication.
```
Comment (mrioan) [inline extensions/servlet/src/main/java/com/stormpath/sdk/servlet/oauth/OAuthTokenResolver.java]: Only coming in a cookie or in headers too? Please clarify that. Thanks

File: extensions/servlet/src/main/java/com/stormpath/sdk/servlet/oauth/impl/AbstractOAuthTokenResolver.java [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+import com.stormpath.sdk.servlet.oauth.OAuthTokenResolver;
+
+import javax.servlet.ServletRequest;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpSession;
+
+/**
+ * @since 1.0.RC10
+ */
+public abstract class AbstractOAuthTokenResolver implements OAuthTokenResolver {
```
Comment (mrioan) [inline extensions/servlet/src/main/java/com/stormpath/sdk/servlet/oauth/impl/AbstractOAuthTokenResolver.java]: @since 1.0.0

File: extensions/servlet/src/main/java/com/stormpath/sdk/servlet/oauth/impl/AccessTokenResolver.java [hunk 12]
```diff
@@ -0,0 +1,21 @@
+package com.stormpath.sdk.servlet.oauth.impl;
+
+/**
+ * Access Token implementation of the {@link com.stormpath.sdk.servlet.oauth.OAuthTokenResolver} interface.
+ *
+ * @since 1.0.RC10
+ */
+public class AccessTokenResolver extends AbstractOAuthTokenResolver {
```
Comment (mrioan) [inline extensions/servlet/src/main/java/com/stormpath/sdk/servlet/oauth/impl/AccessTokenResolver.java]: @since 1.0.0

File: extensions/servlet/src/main/java/com/stormpath/sdk/servlet/oauth/impl/RefreshTokenResolver.java [hunk 13]
```diff
@@ -0,0 +1,21 @@
+package com.stormpath.sdk.servlet.oauth.impl;
+
+/**
+ * Refresh Token implementation of the {@link com.stormpath.sdk.servlet.oauth.OAuthTokenResolver} interface.
+ *
+ * @since 1.0.RC10
```
Comment (mrioan) [inline extensions/servlet/src/main/java/com/stormpath/sdk/servlet/oauth/impl/RefreshTokenResolver.java]: @since 1.0.0

File: extensions/spring/stormpath-spring-webmvc/src/main/java/com/stormpath/spring/config/AbstractStormpathWebMvcConfiguration.java [hunk 14]
```diff
@@ -928,9 +884,9 @@ public MessageSource stormpathSpringMessageSource() {
                 stormpathI18nAlreadyConfigured = true;
             } catch (NoSuchMessageException e) {
                 log.debug(
-                    "Stormpath i18n properties have not been specified during message source configuration.  " +
-                    "Adding these property values as a fallback. Exception for reference (this and the " +
-                    "stack trace can safely be ignored): " + e.getMessage(), e
+                        "Stormpath i18n properties have not been specified during message source configuration.  " +
```
Comment (mrioan) [inline extensions/spring/stormpath-spring-webmvc/src/main/java/com/stormpath/spring/config/AbstractStormpathWebMvcConfiguration.java]: Fix indentation please

File: extensions/spring/stormpath-spring-webmvc/src/main/java/com/stormpath/spring/config/RefreshTokenCookieProperties.java [hunk 15]
```diff
@@ -0,0 +1,59 @@
+package com.stormpath.spring.config;
+
+import com.stormpath.sdk.servlet.config.CookieProperties;
+import org.springframework.beans.factory.annotation.Value;
+
+/**
+ * @since 1.0.RC10
+ */
+public class RefreshTokenCookieProperties implements CookieProperties {
```
Comment (mrioan) [inline extensions/spring/stormpath-spring-webmvc/src/main/java/com/stormpath/spring/config/RefreshTokenCookieProperties.java]: * Please add a line break here
* @since 1.0.0

Comment (mrioan) [inline extensions/spring/stormpath-spring-webmvc/src/main/java/com/stormpath/spring/config/RefreshTokenCookieProperties.java]: * Please add a line break here
* @since 1.0.0

File: extensions/spring/stormpath-spring-webmvc/src/main/java/com/stormpath/spring/config/SpringLoginErrorModelFactory.java [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+
+import javax.servlet.http.HttpServletRequest;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * @since 1.0
+ */
+public class SpringLoginErrorModelFactory implements ErrorModelFactory {
```
Comment (mrioan) [inline extensions/spring/stormpath-spring-webmvc/src/main/java/com/stormpath/spring/config/SpringLoginErrorModelFactory.java]: @since 1.0.0

File: extensions/spring/stormpath-spring-webmvc/src/main/java/com/stormpath/spring/config/StormpathWebMvcConfiguration.java [hunk 17]
```diff
@@ -213,13 +213,13 @@ public String getUnauthorizedUrl() {
...
-            public CookieConfig getAccountCookieConfig() {
-                return stormpathAccountCookieConfig();
+            public CookieConfig getRefreshTokenCookieConfig() {
+                return stormpathRefreshTokenCookieConfig();
             }
 
             @Override
-            public long getAccountJwtTtl() {
-                return accountJwtTtl;
+            public CookieConfig getAccessTokenCookieConfig() {
```
Comment (mrioan) [inline extensions/spring/stormpath-spring-webmvc/src/main/java/com/stormpath/spring/config/StormpathWebMvcConfiguration.java]: Why is the TTL not exposed any longer? Was it never used?

**Comment by mrioan** [comment]
## UAT

- spring-security-webmvc: I do not see the `access_token` or `refresh_token` in cookies after authentication

**Comment by mrioan** [comment]
@jarias Back to you. Thanks!

File: extensions/servlet/src/main/java/com/stormpath/sdk/servlet/config/impl/AbstractCookieConfig.java (hunk 5 above)
Comment (jarias) [inline extensions/servlet/src/main/java/com/stormpath/sdk/servlet/config/impl/AbstractCookieConfig.java]: The thing is that property is loaded at runtime, based on the oauth policy but I going to check if I can avoid setting a value here or put a comment to clarify.

File: extensions/servlet/src/main/java/com/stormpath/sdk/servlet/config/impl/DefaultConfig.java (hunk 6 above)
Comment (jarias) [inline extensions/servlet/src/main/java/com/stormpath/sdk/servlet/config/impl/DefaultConfig.java]: Remove, the cookie ttl or max age is figure out by the Application OAuthPolicy

And for line 84 we don't have an account cookie any more.

File: extensions/servlet/src/main/java/com/stormpath/sdk/servlet/config/impl/DefaultConfig.java (hunk 7 above)
Comment (jarias) [inline extensions/servlet/src/main/java/com/stormpath/sdk/servlet/config/impl/DefaultConfig.java]: Same as above, we are not creating a JWT account anymore is just the access_token, those TTLs where to determine the cookie maxage that according to spec is no determine by the Application OauthPolicy

File: extensions/spring/stormpath-spring-webmvc/src/main/java/com/stormpath/spring/config/StormpathWebMvcConfiguration.java (hunk 17 above)
Comment (jarias) [inline extensions/spring/stormpath-spring-webmvc/src/main/java/com/stormpath/spring/config/StormpathWebMvcConfiguration.java]: It was use in the JWT we use in the Account cookie which no longer used.

**Comment by jarias** [comment]
@mrioan Just update everything according to your review. The spring security was missing all together I thought it used the same authentication mechanism as the the servlet and regular spring.

File: extensions/servlet/src/main/java/com/stormpath/sdk/servlet/oauth/OAuthTokenResolver.java (hunk 10 above)
Comment (jarias) [inline extensions/servlet/src/main/java/com/stormpath/sdk/servlet/oauth/OAuthTokenResolver.java]: Its actually the request attributes, I put that in the JavaDoc

**Comment by dogeared** [comment]
UAT, Round 2:

Positive Test:

1. /login with valid credentials
2. confirm `access_token` and `refresh_token` cookies set
3. Access protected resource
4. /logout
5. confirm `access_token` and `refresh_token` cookies removed

Negative Test:

1. /login with invalid credentials
2. confirm error message

------------------------------------------------------------------------

1. servlet example

    Positive Test: PASS
    Negative Test: PASS

    Notes: If you logout, and then put in the wrong credentials, the view looks like this:

    image

    (This may occur in the current release version)


2. spring webmvc example

    Positive Test: PASS
    Negative Test: PASS

    Note: same note at servlet example

3. spring security webmvc example

    Positive Test: PASS
    Negative Test: **FAIL**

    Notes: while the login was *not* successful as expected for the Negative Test, the error message is not shown.

    image

4. spring boot webmvc example

    Positive Test: PASS
    Negative Test: PASS

    Notes: The issue described in the servlet  example above does **not** occur here.

5. spring security spring boot webmvc example

    Positive Test: PASS
    Negative Test: PASS

6. default spring boot starter example

    Positive Test: PASS
    Negative Test: PASS

**Comment by dogeared** [comment]
@jarias @mrioan based on the above, I've entered two additional issues to address the failures. They are small enough in scope that I am going to merge this PR and we will address the failures separately. The issues are: #647 and #648
